src/main/java/de/bonndan/nivio/input/GroupQueryResolver.java

```diff
@@ -30,5 +31,11 @@ public void process(LandscapeDescription input, LandscapeImpl landscape) {
             }
             groupDescription.getContains().forEach(condition -> group.getItems().addAll(landscape.getItems().query(condition)));
         });
+
+        //cleanup to ensure every items has the group identifier
+        landscape.getItems().stream().forEach(item -> {
+            Group group = landscape.getGroup(item.getGroup()); //if group is empty, COMMON is returned
+            item.setGroup(group.getIdentifier());
```

Comment:
The method landscape.getGroup() has some magic to be non-nullable, so here we ensure the item.group field is set to something that really points to a group. Could make an example how Landscape.updateGroupsOnItems might be used?

Reply:
Yes, I saw that it uses that `Group.COMMON` and thanks for your explanation! I am still concerned about the mutability of `Item` and of the way we expose the internals of the objects.

So if your example means that you would agree to a refactoring (because you call it `Landscape.updateGroupsOnItems`): yes, please and I'll do the refactoring if you like.

I am still wondering why the identifier can not be set on the item right from the beginning? Why exactly isn't it possible to have such (important) fields immutable right from the beginning when we create an instance of the object?

Reply:
You are right about the immutability, although some things cannot be set right from the start. And if we're strict about immutability and side effect free programming, we end in reference update hell (this is still a graph with lots of references).

My idea was that all graph components are value objects without much business logic. The logic was meant to reside in the resolvers as far as possible. So we have small units (resolvers) for one dedicated purpose and small value objects. 
Obviously the landscape.getGroup() method does too much, and I'd rather go the other way and refactor even more logic out of the graph components.

Reply:
Ok, I get that about the resolvers and the value object - that approach is fine. Since I lack the overview of how the value objects are used, I was concerned about the objects having an invalid state (in a field such as an identifier). If only a designated resolver manages them, and the logic resides inside of that - that should be fine.